safehttp/plugins/xsrf/xsrf.go

```diff
@@ -18,146 +18,229 @@
+	c, err := r.Cookie(i.cookieIDKey)
+	if err != nil {
+		buf := make([]byte, 20)
+		if _, err := rand.Read(buf); err != nil {
+			return fmt.Errorf("crypto/rand.Read: %v", err)
 		}
+		c = safehttp.NewCookie(i.cookieIDKey, base64.StdEncoding.EncodeToString(buf))
+		c.SetSameSite(safehttp.SameSiteStrictMode)
 
-		if ok := xsrftoken.Valid(tok, it.SecretAppKey, cookieID.Value(), actionID); !ok {
-			return w.WriteError(safehttp.StatusForbidden)
+		if err := w.SetCookie(c); err != nil {
+			return err
 		}
 	}
 
-	tok := xsrftoken.Generate(it.SecretAppKey, cookieID.Value(), actionID)
-	r.SetContext(context.WithValue(r.Context(), tokenCtxKey{}, tok))
-	return safehttp.NotWritten()
-}
+	tok := xsrftoken.Generate(i.secretAppKey, c.Value(), r.URL.Path())
 
-// Commit adds the XSRF token corresponding to the safehttp.TemplateResponse
-// with key "XSRFToken". The token corresponds to the user information found in
-// the request.
-func (it *Interceptor) Commit(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, resp safehttp.Response, _ safehttp.InterceptorConfig) safehttp.Result {
 	tmplResp, ok := resp.(safehttp.TemplateResponse)
 	if !ok {
-		return safehttp.NotWritten()
-	}
-
-	tok, err := Token(r)
-	if err != nil {
-		// The token should have been added in the Before stage and if that is
-		// not the case, a server misconfiguration occured.
-		return w.WriteError(safehttp.StatusInternalServerError)
+		return nil
 	}
 
 	// TODO(maramihali@): Change the key when function names are exported by
 	// htmlinject
 	// TODO: what should happen if the XSRFToken key is not present in the
 	// tr.FuncMap?
 	tmplResp.FuncMap["XSRFToken"] = func() string { return tok }
-	return safehttp.NotWritten()
+	return nil
+}
+
+type angularChecker struct {
+	tokenCookieName string
+	tokenHeaderName string
+}
+
+func (c angularChecker) Retrieve(r *safehttp.IncomingRequest) (token, userID, actionID string, code safehttp.StatusCode) {
+	cookie, err := r.Cookie(c.tokenCookieName)
+	if err != nil {
+		code = safehttp.StatusForbidden
+		return
+	}
+	token = r.Header.Get(c.tokenHeaderName)
+	if token == "" || token != cookie.Value() {
+		code = safehttp.StatusUnauthorized
+		return
+	}
+	return
+}
+
+func (c angularChecker) Validate(_, _, _ string) safehttp.StatusCode {
+	return safehttp.StatusOK
+}
+
+type angularInjector struct {
+	tokenCookieName string
+}
+
+func (i angularInjector) Inject(resp safehttp.Response, w *safehttp.ResponseWriter, r *safehttp.IncomingRequest) error {
+	c, err := r.Cookie(i.tokenCookieName)
+	if err == nil {
+		return nil
+	}
+	tok := make([]byte, 20)
+	if _, err := rand.Read(tok); err != nil {
+		return fmt.Errorf("crypto/rand.Read: %v", err)
+	}
+	c = safehttp.NewCookie(i.tokenCookieName, base64.StdEncoding.EncodeToString(tok))
+
+	c.SetSameSite(safehttp.SameSiteStrictMode)
+	c.SetPath("/")
+	// Set the duration of the token cookie to 24 hours.
+	c.SetMaxAge(86400)
```

Comment:
We should be coherent and expire the other XSRF tokens in the same amount of time.

I think 24 hours is a bit tight, what about setting this and the other to a week?